dandi/utils.py

```diff
@@ -834,3 +835,32 @@ def _prepare_path_parts(paths: Iterable[str | PurePath]) -> list[tuple[str, ...]
 
 def _starts_with(t: tuple[str, ...], prefix: tuple[str, ...]) -> bool:
     return t[: len(prefix)] == prefix
+
+
+def pre_upload_size_check(path: Path) -> int:
+    # If the filesystem reports a size of zero for a file we're about to
+    # upload, double-check the size in case we're on a flaky NFS system.
+    for naptime in [0] + [0.1] * 19:
+        sleep(naptime)
+        size = path.stat().st_size
+        if size != 0:
+            return size
+    return size
+
+
+def post_upload_size_check(path: Path, pre_check_size: int, erroring: bool) -> None:
+    # More checks for NFS flakiness
+    size = path.stat().st_size
+    if size != pre_check_size:
+        if erroring:
+            lgr.error(
+                "Size of %s was %d at start of upload but is now %d after upload",
+                path,
+                pre_check_size,
+                size,
+            )
+        else:
+            raise RuntimeError(
+                f"Size of {path} was {pre_check_size} at start of upload but is"
+                f" now {size} after upload"
+            )
```

Comment:
Now that we know it is safe for logger to just have a string possibly with `%`s, let's just please anti-duplication gods with:

```suggestion
        msg = f"Size of {path} was {pre_check_size} at start of upload but is" \
            f" now {size} after upload"
        if erroring:
            lgr.error(msg)
        else:
            raise RuntimeError(msg)
```

?

Reply:
having a little tiny test to cover this codepath would be also great.

Reply:
@yarikoptic How exactly am I supposed to change the filesize in the middle of an upload for a test?

Reply:
unittest at the level of `post_upload_size_check` function itself -- just provide a wrong filesize thus to trigger logging or exception

Reply:
if you want for the entire upload -- just as a thought excercise, I thought it could be via instrumentation of that `super_len` function so it changes the file after returning the size ;-)